File: Level - 2/Divide and conquer/Sqrt(x).cpp
class Solution {
public:
    int mySqrt(int x) {
        //OM GAN GANAPATHAYE NAMO NAMAH 
        //JAI SHRI RAM 
        //JAI BAJRANGBALI 
        //AMME NARAYANA, DEVI NARAYANA, LAKSHMI NARAYANA, BHADRE NARAYANA
        if(x == 0)
            return 0;
        else if(x == 1)
            return 1;
        int l = 0 , r = x;
        int res = -1;
        while(l <= r)
        {
            int mid = (int)((l+r)/2);
            if(mid == x/mid)
            {
                return mid;
            }
            else if(mid < x/mid)
            {
                res = mid;
                l = mid + 1;
            }
            else
            {
                r = mid - 1;
            }
        }
        return res;
    }
};

File: DSA Revision/Sid/Binary Search/Sqrt(x).java
class Solution {
    //OM GAN GANAPATHAYE NAMO NAMAH 
    //JAI SHRI RAM 
    //JAI BAJRANGBALI 
    //AMME NARAYANA, DEVI NARAYANA, LAKSHMI NARAYANA, BHADRE NARAYANA
    public int mySqrt(int x) {
        if(x == 0 || x == 1)
            return x;
        int l = 1, r = x - 1;
        int ans = -1;
        while(r >= l)
        {
            int mid = (r+l)/2;
            if(mid <= (x/mid))
            {
                ans = mid; 
                l = mid + 1;
            }
            else
                r = mid - 1;
        }
        return ans;
    }
}
